readthedocs/projects/tasks/builds.py

```diff
@@ -0,0 +1,1020 @@
+            python_env=self.python_env,
+        )
+        return html_builder.get_final_doctype()
+
+    def build_docs_search(self):
+        """
+        Build search data.
+
+        .. note::
+           For MkDocs search is indexed from its ``html`` artifacts.
+           And in sphinx is run using the rtd-sphinx-extension.
+        """
+        return self.is_type_sphinx()
+
+    def build_docs_localmedia(self):
+        """Get local media files with separate build."""
+        if (
+            'htmlzip' not in self.config.formats or
+            self.version.type == EXTERNAL
+        ):
+            return False
+        # We don't generate a zip for mkdocs currently.
+        if self.is_type_sphinx():
+            return self.build_docs_class('sphinx_singlehtmllocalmedia')
+        return False
+
+    def build_docs_pdf(self):
+        """Build PDF docs."""
+        if 'pdf' not in self.config.formats or self.version.type == EXTERNAL:
+            return False
+        # Mkdocs has no pdf generation currently.
+        if self.is_type_sphinx():
+            return self.build_docs_class('sphinx_pdf')
+        return False
+
+    def build_docs_epub(self):
+        """Build ePub docs."""
+        if 'epub' not in self.config.formats or self.version.type == EXTERNAL:
+            return False
+        # Mkdocs has no epub generation currently.
+        if self.is_type_sphinx():
+            return self.build_docs_class('sphinx_epub')
+        return False
+
+    def build_docs_class(self, builder_class):
+        """
+        Build docs with additional doc backends.
+
+        These steps are not necessarily required for the build to halt, so we
+        only raise a warning exception here. A hard error will halt the build
+        process.
+        """
+        builder = get_builder_class(builder_class)(
+            self.build_env,
+            python_env=self.python_env,
+        )
+        success = builder.build()
+        builder.move()
+        return success
+
+    def send_notifications(self, version_pk, build_pk, event):
+        """Send notifications to all subscribers of `event`."""
+        # Try to infer the version type if we can
+        # before creating a task.
+        if not self.version or self.version.type != EXTERNAL:
+            build_tasks.send_build_notifications.delay(
+                version_pk=version_pk,
+                build_pk=build_pk,
+                event=event,
+            )
+
+    def is_type_sphinx(self):
+        """Is documentation type Sphinx."""
+        return 'sphinx' in self.config.doctype
+
+
+# TODO: rename this task to `build_task`
+@app.task(
+    base=UpdateDocsTask,
+    bind=True,
+)
+def update_docs_task(self, *args, **kwargs):
+    # TODO: `build_pk` is a mandatory attribute. We cannot trigger this task
+    # without a build already created in the database.
+
+    # HACK: just for now while we are developing the code
+    self.request.kwargs['version_pk'] = args[0]
```

Comment:
Just a note here so this doesn't get lost, this todo sounds important.

Reply:
I did this temporarily. Basically, I want to have required arguments _being required_ instead of optional keyword arguments. They are mainly `version_pk` and `build_pk`. So, to avoid changing the signature of the task and breaking a lot of code immediately when I started the refactor, I did this hack for now.